runtime/lua/vim/filetype/detect.lua

```diff
@@ -1204,4 +1286,269 @@ end
+  ['^SNNS network definition file'] = 'snnsnet',
+  ['^SNNS pattern definition file'] = 'snnspat',
+  ['^SNNS result file'] = 'snnsres',
+  ['&%%.-[Vv]irata'] = { 'virata', { start_lnum = 1, end_lnum = 5 } },
+  ['[0-9:.]* *execve%('] = 'strace',
+  ['^__libc_start_main'] = 'strace',
+  -- VSE JCL
+  ['^\\* $$ JOB\\>'] = { 'vsejcl', { vim_regex = true } },
+  ['^// *JOB\\>'] = { 'vsejcl', { vim_regex = true } },
+  -- TAK and SINDA
+  ['K & K  Associates'] = { 'takout', { start_lnum = 4 } },
+  ['TAK 2000'] = { 'takout', { start_lnum = 2 } },
+  ['S Y S T E M S   I M P R O V E D '] = { 'syndaout', { start_lnum = 3 } },
+  ['Run Date: '] = { 'takcmp', { start_lnum = 6 } },
+  ['Node    File  1'] = { 'sindacmp', { start_lnum = 9 } },
+  function(contents)
+    require('vim.filetype.detect').dns_zone(contents)
+  end,
+  -- Valgrind
+  ['^==%d+== valgrind'] = 'valgrind',
+  ['^==%d+== Using valgrind'] = { 'valgrind', { start_lnum = 3 } },
+  -- Go docs
+  ['PACKAGE DOCUMENTATION$'] = 'godoc',
+  -- Renderman Interface Bytestream
+  ['^##RenderMan'] = 'rib',
+  -- Scheme scripts
+  ['exec%s%+%S*scheme'] = { 'scheme', { start_lnum = 1, end_lnum = 2 } },
+  -- Git output
+  ['^\\(commit\\|tree\\|object\\) \\x\\{40,\\}\\>\\|^tag \\S\\+$'] = { 'git', { vim_regex = true } },
+  function(lines)
+    -- Gprof (gnu profiler)
+    if lines[1] == 'Flat profile:' and lines[2] == '' and lines[3]:find('^Each sample counts as .* seconds%.$') then
+      return 'gprof'
+    end
+  end,
+  -- Erlang terms
+  -- (See also: http://www.gnu.org/software/emacs/manual/html_node/emacs/Choosing-Modes.html#Choosing-Modes)
+  ['%-%*%-.*erlang.*%-%*%-'] = { 'erlang', { ignore_case = true } },
+  -- YAML
+  ['^&YAML'] = 'yaml',
+  -- MikroTik RouterOS script
+  ['^#.*by RouterOS'] = 'routeros',
+  -- Sed scripts
+  -- #ncomment is allowed but most likely a false positive so require a space before any trailing comment text
+  ['^#n%s'] = 'sed',
+  ['^#n$'] = 'sed',
+}
+
+---@private
+-- File does not start with "#!".
+local function match_from_text(contents, path)
+  if contents[1]:find('^:$') then
+    -- Bourne-like shell scripts: sh ksh bash bash2
+    return M.sh(path, contents)
+  elseif matchregex('\n' .. table.concat(contents, '\n'), [[\n\s*emulate\s\+\%(-[LR]\s\+\)\=[ckz]\=sh\>]]) then
+    -- Z shell scripts
+    return 'zsh'
+  end
+
+  for k, v in pairs(patterns_text) do
+    if type(v) == 'string' then
+      -- Check the first line only
+      if contents[1]:find(k) then
+        return v
+      end
+    elseif type(v) == 'function' then
+      -- If filetype detection fails, continue with the next pattern
+      local ok, ft = pcall(v, contents)
+      if ok and ft then
+        return ft
+      end
+    else
+      local opts = type(v) == 'table' and v[2] or {}
+      if opts.start_lnum then
+        assert(not opts.ignore_case, 'ignore_case=true is ignored when start_lnum is also present, needs refactor')
+        for i = opts.start_lnum, opts.end_lnum do
```

Comment:
```suggestion
        for i = opts.start_lnum, opts.end_lnum or -1 do
```

Reply:
Won't the fix in your commit run into problems if only start_lnum is specified (which is the case for a few patterns?

Maybe it's more robust to require specifying the end_lnum even if it's -1?

Reply:
I changed it so that we do not loop when only `start_lnum` is specified, instead we should jump into the else branch. Maybe I misunderstood something though. I think defaulting `end_lnum` to -1 is not correct because this would try to access line -1.

Reply:
Right, I just saw that (was on mobile before).

`-1` means "last line", so this would loop through all lines -- but it looks like that was not the intention, so your solution is fine.

Reply:
No, sorry, this still breaks. You need to put a  `finish` at the top of `scripts.vim` to make sure that `test_filetype_lua` exercises your code.

Reply:
I really think `end_lnum` should default to `-1` -- otherwise those tests don't pass. My fix passes the tests, while yours doesn't.

Reply:
> You need to put a finish at the top of scripts.vim to make sure that test_filetype_lua exercises your code.

Sorry, I am not sure I can follow (apologies for any stupid questions, I think it'd be best if I took a break :sweat_smile:). I think this is already present in `scripts.vim`, no? Btw, I noticed earlier that it must be `exists("g:do_filetype_lua")` instead of `exists(g:do_filetype_lua)`.

Reply:
Yeah, it's late here as well.

The issue is that our `test_filetype_lua` just doesn't work here (it's fundamentally flawed, which is not your fault), since the `scripts.vim` is sourced _before_ the test script sets the variable. So `scripts.vim` is _always_ executed and sets the right filetype even if the `match_content` function fails. The only way to make sure you're actually testing the new functions here is to basically delete `scripts.vim` (locally, since we of course can't do it in the repo yet).

Reply:
(I meant an _unconditional_ `finish` -- no `if`s and buts!)